### DISCUSSION TITLE
Things work in progress

### DISCUSSION THREAD

**Discussion by gtxzsxxk** [gtxzsxxk/duo-buildroot#18] [Q&A]
I need to submit our changes to the mainline within the next week. Let me check the things we need to do.

**Comment by gtxzsxxk** [comment]
1. add two configs 

```bash
configs/milkv_duo_musl_riscv64_defconfig # which will leave some memory for the camera use. Only 28MiB will be free.
configs/milkv_duo_musl_riscv64_64mb_defconfig # which will have all the 64MiB free memory
```

**Reply by gtxzsxxk** [reply]
done

**Reply by shodan8192** [reply]
I like the way you've done support for free memory selection. In addition, I think that also kernel should have `linux-64mb` patchset specified as `BR2_LINUX_KERNEL_PATCH`, **without** those lines in `0005-feat-add-milkv-duo-kernel-defconfig.patch` :
```
+CONFIG_ION=y
+CONFIG_ION_SYSTEM_HEAP=y
+CONFIG_ION_CARVEOUT_HEAP=y
+CONFIG_ION_CMA_HEAP=y
```
to remove ION support completely.

**Reply by gtxzsxxk** [reply]
Thanks! Let me do this! I'd like to directly work here:
https://github.com/milkv-duo/milkv-duo-buildroot

And I cannot handle the settings on that forked repo to open the discussion or issue because it does not have a settings tab at all. So let us keep discussing here.

**Reply by shodan8192** [reply]
OK, that's fine. I have some more ideas, so the future PRs should I open on https://github.com/milkv-duo/milkv-duo-buildroot instead of here ?

**Reply by gtxzsxxk** [reply]
Yes, feel free and please. And I plan to submit our changes at most the next weekend.

**Reply by gtxzsxxk** [reply]
```
+CONFIG_ION=y
+CONFIG_ION_SYSTEM_HEAP=y
+CONFIG_ION_CARVEOUT_HEAP=y
+CONFIG_ION_CMA_HEAP=y
```

Hello Emil, I am sorry that could we ignore these lines? These changes seems minor and however delete them like what I did in u-boot might bring more files and changes, which might cause trouble when submitting.

**Comment by gtxzsxxk** [comment]
2. add hash for our packages

**Comment by gtxzsxxk** [comment]
3. sync with the master branch. All changes here since now will be finally moved to https://github.com/milkv-duo/milkv-duo-buildroot

**Reply by gtxzsxxk** [reply]
done

**Comment by gtxzsxxk** [comment]
4. Maybe we should do some deletion. I just found that it seems that our changes are much more heavier than other boards. I am afraid that our submit will be requested changes.

**Reply by shodan8192** [reply]
Linux and u-boot patches seems to be heaviest here. Maybe we should have repos holding those already patched (like fsbl), and leave here only patches for memory selecting ?

**Reply by gtxzsxxk** [reply]
Because they haven't been merged into the mainline currently. Maybe in the future it can be much more brief.

**Reply by gtxzsxxk** [reply]
I took a look at the ARM boards. They usually do not have any patches. I think it is because they had developed linux for years and all code are merged into mainline so as nothing need to be fixed by patches.

Maybe I should send an email asking them, to test their attitude on this.

**Reply by shodan8192** [reply]
You've right about this. So maybe separate repos for u-boot and kernel indeed could be solution for now ?

**Reply by gtxzsxxk** [reply]
Certainly. But they must be official. Take a look at `configs/odroidxu4_defconfig`.

```
BR2_LINUX_KERNEL=y
BR2_LINUX_KERNEL_CUSTOM_GIT=y
BR2_LINUX_KERNEL_CUSTOM_REPO_URL="https://github.com/hardkernel/linux.git"
BR2_LINUX_KERNEL_CUSTOM_REPO_VERSION="4.14.16-104"
BR2_LINUX_KERNEL_DEFCONFIG="odroidxu4"
BR2_LINUX_KERNEL_DTS_SUPPORT=y
BR2_LINUX_KERNEL_INTREE_DTS_NAME="exynos5422-odroidxu4"
BR2_TARGET_ROOTFS_EXT2=y
BR2_TARGET_ROOTFS_EXT2_4=y
```

**Reply by gtxzsxxk** [reply]
I have never thought this possible, until I have the access to the milkv-duo's github organization. Haha.

You are right, that is excellently good. And by this I can remove the ugly commit id. That's too great.

**Reply by shodan8192** [reply]
So, is this doable at https://github.com/milkv-duo/linux.git ?

**Reply by gtxzsxxk** [reply]
Yes. I need to ask my supervisor and they have high possibility of agreement.

**Reply by gtxzsxxk** [reply]
OK I got the permission. Let's do this

**Reply by shodan8192** [reply]
That is a great news !

**Reply by gtxzsxxk** [reply]
Done. Now u-boot and linux have been directly our repos. There are no more uboot and linux patches now.

**Reply by shodan8192** [reply]
Good job ! It looks clean and simple now. I'll try it tomorrow.

**Reply by shodan8192** [reply]
Now, when we no need tradeoffs also kernel should have non-ION config for 64MB I think.

**Reply by gtxzsxxk** [reply]
OK. I will open a new tag about non-ION on the kernel.

**Reply by shodan8192** [reply]
You don't have to. Just add another config in https://github.com/milkv-duo/milkv-duo-linux/tree/linux-5.10.4/arch/riscv/configs, and point to it in buildroot config

**Reply by gtxzsxxk** [reply]
Sorry for that.. I will do that right now.

**Reply by shodan8192** [reply]
Maybe also `milkv-duo-u-boot` can have a build switch (`FREE_RAM_SIZE`) like `milkv-duo-firmware-fsbl` does, to control memory map ? Need of two tags can be avoided in this manner.
